### PR TITLE
Add WordSearch problem solution and Dictionary class

### PR DESCRIPTION
**Pull request by mernarezk**

### CODE REVIEW

File: mernarezk/assignment3/WordSearch.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+			}
+		}
+		return result;
+	}
+
+	private void explore(char grid[][], Dictionary dictionary, Set<String> result, boolean visited[][], int i, int j,
+			String word) {
+		if (dictionary.isWord(word)) {
+			result.add(word);
+			return;
```
Comment (imorlowska) [inline mernarezk/assignment3/WordSearch.java]: What if you had a grid ['c', 'a', 'r', 'd'] and words "car" and "card"? :)

File: mernarezk/assignment3/WordSearch.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		} else {
```
Comment (imorlowska) [inline mernarezk/assignment3/WordSearch.java]: What if you had a grid [a, a, a, a, a, a, a, ... , a] (a LOT of 'a's) and the words "ab" and "aab"? When would this search stop starting from the first character in the grid?

File: mernarezk/assignment3/WordSearchTest.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+		dictionary.addWord("cat");
+		
+		Set<String> resultContainsTwoWords = tester.searchWords(grid, dictionary);
+		Set<String> resultWithLetterUsedOnce = tester.searchWords(grid, dictionary2);
+		assertEquals(resultWithLetterUsedOnce.contains("ccar"),false);
+		assertEquals(resultContainsTwoWords.contains("car"),true);
+		assertEquals(resultContainsTwoWords.contains("cat"),true);
+		assertEquals(resultContainsTwoWords.size(),2);
+
+
```
Comment (imorlowska) [inline mernarezk/assignment3/WordSearchTest.java]: Please remove extra empty lines.

File: mernarezk/assignment3/Dictionary.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	public void addWord(String s) {
+		generatePrefixes(s);
+		words.add(s);
+	}
+
+	private Set<String> generatePrefixes(String s) {
+		StringBuilder str = new StringBuilder("");
+		for (int i = 0; i < s.length(); i++) {
+			str.append(s.charAt(i));
+			prefixes.add(str.toString());
```
Comment (imorlowska) [inline mernarezk/assignment3/Dictionary.java]: It could also be:

```
for (int i = 1; i <= s.length(); i++) {
    prefixes.add(str.substring(0, i));
}
```

File: mernarezk/assignment3/WordSearchTest.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+
+import java.util.Set;
+
+import org.junit.jupiter.api.Test;
+
+class WordSearchTest {
+	
+	@Test
```
Comment (imorlowska) [inline mernarezk/assignment3/WordSearchTest.java]: How do you feel about adding some more tests? Like using a grid that you design, with words of your choosing and checking for different edge cases like empty grid, empty dictionary, reusing letters, words on a diagonal, words that could be created (but shouldn't) if skipping cells was allowed etc.
You could also unit test some of your helper methods, or the Dictionary class methods.

File: mernarezk/assignment3/Dictionary.java (hunk 4 above)
Comment (mernarezk) [reply mernarezk/assignment3/Dictionary.java]: oh yes but I prefer using stringbuilder because it takes less time to append to a string O(n) while substring takes O(n^2) since strings are immutable in java.

File: mernarezk/assignment3/WordSearch.java (hunk 1 above)
Comment (mernarezk) [reply mernarezk/assignment3/WordSearch.java]: aagghh yes it was a bug it used to return just car. fixed it.

File: mernarezk/assignment3/WordSearch.java (hunk 2 above)
Comment (mernarezk) [reply mernarezk/assignment3/WordSearch.java]: oh I removed the else because I figured out it was a mistake but I still think it will go over all the a's since I have two outer loops that try out all possible start points from the grid

File: mernarezk/assignment3/WordSearchTest.java (hunk 3 above)
Comment (mernarezk) [reply mernarezk/assignment3/WordSearchTest.java]: done

File: mernarezk/assignment3/WordSearchTest.java (hunk 5 above)
Comment (imorlowska) [reply mernarezk/assignment3/WordSearchTest.java]: Very nice! Just for the future I'd say perhaps it would be nice to divide such a method into a few smaller tests, but all looks good! :)
